cassiopeia/plotting/itol_utilities.py

```diff
@@ -0,0 +1,623 @@
+    temporary_directory = tempfile.mkdtemp()
+
+    if os.path.exists(os.path.expanduser(itol_config)):
+        
+        config = configparser.ConfigParser()
+        with open(os.path.expanduser(itol_config), "r") as f:
+            config_string = f.read()
+        config.read_string(config_string)
+        if api_key is None:
+            api_key = config["DEFAULT"]["api_key"]
+        if project_name is None:
+            project_name = config["DEFAULT"]["project_name"]
+
+    if api_key is None or project_name is None:
+        raise iTOLError(
+            "Specify an api_key and project_name, or a valid iTOL "
+            "config file."
+        )
+
+    with open(os.path.join(temporary_directory, "tree_to_plot.tree"), "w") as f:
+        f.write(cassiopeia_tree.get_newick())
+
+    file_format = export_filepath.split("/")[-1].split(".")[-1]
+
+    if file_format not in ["png", "svg", "eps", "ps", "pdf"]:
+        raise iTOLError(
+            "File format must be one of " "'png', 'svg', 'eps', 'ps', 'pdf']"
+        )
+
+    itol_uploader = Itol()
+    itol_uploader.add_file(os.path.join(temporary_directory, "tree_to_plot.tree"))
+
+    files = []
+    if allele_table is not None:
+        files += create_indel_heatmap(
+            allele_table,
+            cassiopeia_tree,
+            f"{tree_name}.allele",
+            temporary_directory,
+            indel_colors,
+            indel_priors,
+            random_state,
+        )
+
+    for meta_item in meta_data:
+        if meta_item not in cassiopeia_tree.cell_meta.columns:
+            raise iTOLError("Meta data item not in CassiopeiaTree cell meta.")
+
+        values = cassiopeia_tree.cell_meta[meta_item]
+
+        if pd.api.types.is_numeric_dtype(values):
+            files.append(create_gradient_from_df(
+                values, cassiopeia_tree, f"{tree_name}.{meta_item}", temporary_directory,
+            ))
+
+        if pd.api.types.is_string_dtype(values):
+            colors = palette[: len(values.unique())]
+            colors = [hex_to_rgb(color) for color in colors]
+            colormap = dict(zip(np.unique(values), colors))
+
+            files.append(create_colorbar(
+                values,
+                cassiopeia_tree,
+                colormap,
+                f"{tree_name}.{meta_item}",
+                temporary_directory,
+                create_legend=include_legend,
+            ))
+
+    for _file in files:
+        itol_uploader.add_file(_file)
+
+    itol_uploader.params["treeName"] = tree_name
+    itol_uploader.params["APIkey"] = api_key
+    itol_uploader.params["projectName"] = project_name
+
+    good_upload = itol_uploader.upload()
+    if not good_upload:
+        raise iTOLError(itol_uploader.comm.upload_output)
+
+    print("iTOL output: " + str(itol_uploader.comm.upload_output))
+    print("Tree Web Page URL: " + itol_uploader.get_webpage())
+    print("Warnings: " + str(itol_uploader.comm.warnings))
```

Comment:
Should we toggle these print statements with a verbosity flag?

Reply:
done.